utility/app.go

```diff
@@ -0,0 +1,419 @@
+	}
+	// ensure app doesn't already exist
+	exists, err := u.GetAppExists(publicKey.Address())
+	if err != nil {
+		return err
+	}
+	if exists {
+		return types.ErrAlreadyExists()
+	}
+	// insert the app structure
+	if err := u.InsertApplication(publicKey.Address(), message.PublicKey, message.OutputAddress, maxRelays, message.Amount, message.Chains); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u *UtilityContext) HandleMessageEditStakeApp(message *utilTypes.MessageEditStakeApp) types.Error {
+	exists, err := u.GetAppExists(message.Address)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return types.ErrNotExists()
+	}
+	amountToAdd, err := types.StringToBigInt(message.AmountToAdd)
+	if err != nil {
+		return err
+	}
+	// ensure signer has sufficient funding for the stake
+	signerAccountAmount, err := u.GetAccountAmount(message.Signer)
+	if err != nil {
+		return err
+	}
+	signerAccountAmount.Sub(signerAccountAmount, amountToAdd)
+	if signerAccountAmount.Sign() == -1 {
+		return types.ErrInsufficientAmountError()
+	}
+	maxChains, err := u.GetAppMaxChains()
+	if err != nil {
+		return err
+	}
+	// validate chains
+	if len(message.Chains) > maxChains {
+		return types.ErrMaxChains(maxChains)
+	}
+	// update account amount
+	if err := u.SetAccount(message.Signer, signerAccountAmount); err != nil {
+		return err
+	}
+	// move funds from account to pool
+	if err := u.AddPoolAmount(utilTypes.AppStakePoolName, amountToAdd); err != nil {
+		return err
+	}
+	// calculate maximum relays from stake amount
+	maxRelaysToAdd, err := u.CalculateAppRelays(message.AmountToAdd)
+	if err != nil {
+		return err
+	}
+	// insert the app structure
+	if err := u.UpdateApplication(message.Address, maxRelaysToAdd, message.AmountToAdd, message.Chains); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u *UtilityContext) HandleMessageUnstakeApp(message *utilTypes.MessageUnstakeApp) types.Error {
+	status, err := u.GetAppStatus(message.Address)
+	if err != nil {
+		return err
+	}
+	// validate is staked
+	if status != utilTypes.StakedStatus {
+		return types.ErrInvalidStatus(status, utilTypes.StakedStatus)
+	}
+	unstakingHeight, err := u.CalculateAppUnstakingHeight()
+	if err != nil {
+		return err
+	}
+	if err := u.SetAppUnstakingHeightAndStatus(message.Address, unstakingHeight, utilTypes.UnstakingStatus); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u *UtilityContext) UnstakeAppsThatAreReady() types.Error {
+	appsReadyToUnstake, err := u.GetAppsReadyToUnstake()
+	if err != nil {
+		return err
+	}
+	for _, app := range appsReadyToUnstake {
+		if err := u.SubPoolAmount(utilTypes.AppStakePoolName, app.GetStakeAmount()); err != nil {
+			return err
+		}
+		if err := u.AddAccountAmountString(app.GetOutputAddress(), app.GetStakeAmount()); err != nil {
+			return err
+		}
+		if err := u.DeleteApplication(app.GetAddress()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (u *UtilityContext) BeginUnstakingMaxPausedApps() types.Error {
```

Comment:
Because they aren't actually unstaking - they are just beginning to unstake (21 day process). See how the explicit naming makes you dig a bit deeper :)

Reply:
Amazing.